### ISSUE TITLE
ERROR: LoadError: MethodError: no method matching (::Colon)(::Nothing, ::Int64, ::Int64)

### ISSUE THREAD

**Issue by fbanning** [JuliaDynamics/InteractiveDynamics.jl#12]
I am testing the package with a ContinuousSpace model and want to plot some data in the interactive widget.
So I define the following model which I've simplified a bit:

``` julia
mutable struct MyAgent <: AbstractAgent
    id::Int
    pos::Tuple{Float64, Float64}
    vel::Tuple{Float64, Float64}
    money::Float64
    transfer::Float64
end

# model, stepping functions and definitions of ac and am go here

adata = [(:money, sum), (:transfer, sum)]
alabels = ["Aggregate money", "Aggregate transfers"]

interactive_abm(
    model, agent_step!, dummystep, params; 
    ac = ac, as = 0.02, am = am,
    adata = adata, alabels = alabels
)
```

When I execute this, I get the following error message:
```
ERROR: LoadError: MethodError: no method matching (::Colon)(::Nothing, ::Int64, ::Int64)
```
What exactly went wrong here?
There's a very long stacktrace which I'm so far unable to see through.
The interactive_abm function starts perfectly fine when I don't define `adata` and `alabels` so I don't expect it to be a problem with the model itself.

**Comment by Datseris** [comment]
First of all, thanks a lot for taking the time to test this library and report errors! Super appreciated!

Now, to help you, I will need a MWE (Minimal Working Example), or, if you don't have the time to produce one, the full code that you run and get the error. 

The full stacktrace is extremely necessary always, even if you can't read it. It tells me which is the function that errors, and whether this comes from my module, another module, or even your own code.

**Comment by Datseris** [comment]
> The interactive_abm function starts perfectly fine when I don't define adata and alabels so I don't expect it to be a problem with the model itself.

It could be that your data aggregation is ill defined. Are agents killed to the point that noone remains? You can't aggregate over that.

**Comment by fbanning** [comment]
> Are agents killed to the point that noone remains?

No, none of the agents die.

> data aggregation is ill defined.

I thought my test was simple enough so that there is no chance for a custom aggregation function to mess things up. It is not a problem to aggregate this exact data in a normal `run!` and collect it in a df. The problem only occurs when running via the interactive app.

Now I've run this again today and what can I tell you, the app started right away and let me run the model once (yay, I guess). After that I clicked on setup/update to reset the model to its initial state but that again threw an error. It's most likely the same error as before but I'm not 100% certain.
Here's the stacktrace:

``` julia
Error in c callback: 
MethodError: no method matching (::Colon)(::Nothing, ::Int64, ::Int64)
Closest candidates are:
  Any(::T, ::T, ::T) where T<:Real at range.jl:22
  Any(::T, ::Any, ::T) where T<:Real at range.jl:41
  Any(::A, ::Any, ::C) where {A<:Real, C<:Real} at range.jl:10
  ...
Stacktrace:
 [1] locateticks(::Float32, ::Float32, ::Int64, ::Bool, ::Int64) at C:\Users\fbanning\.julia\packages\MakieLayout\3a66O\src\ticklocators\linear.jl:101
 [2] locateticks at C:\Users\fbanning\.julia\packages\MakieLayout\3a66O\src\ticklocators\linear.jl:77 [inlined]
 [3] get_tickvalues at C:\Users\fbanning\.julia\packages\MakieLayout\3a66O\src\lineaxis.jl:326 [inlined]
 [4] get_tickvalues at C:\Users\fbanning\.julia\packages\MakieLayout\3a66O\src\lineaxis.jl:318 [inlined]
 [5] (::MakieLayout.var"#104#125")(::Tuple{Float32,Tuple{Float32,Float32},Bool}, ::Tuple{Float32,Float32}, ::AbstractPlotting.Automatic) at C:\Users\fbanning\.julia\packages\MakieLayout\3a66O\src\lineaxis.jl:144   
 [6] (::Observables.MapUpdater{MakieLayout.var"#104#125",Array{Float64,1}})(::Tuple{Float32,Tuple{Float32,Float32},Bool}, ::Vararg{Any,N} where N) at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:241
 [7] (::Observables.OnUpdate{Observables.MapUpdater{MakieLayout.var"#104#125",Array{Float64,1}},Tuple{Observable{Tuple{Float32,Tuple{Float32,Float32},Bool}},Observable{Any},Observable{Any}}})(::Tuple{Float32,Float32}) at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:218
 [8] setindex!(::Observable{Any}, ::Tuple{Float32,Float32}; notify::Observables.var"#6#8") at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:130
 [9] setindex! at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:126 [inlined]
 [10] (::Observables.var"#3#4"{Any,Observable{Any}})(::Tuple{Float32,Float32}) at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:47
 [11] #invokelatest#1 at .\essentials.jl:710 [inlined]
 [12] invokelatest at .\essentials.jl:709 [inlined]
 [13] setindex!(::Observable{Tuple{Float32,Float32}}, ::Tuple{Float32,Float32}; notify::Observables.var"#6#8") at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:132
 [14] setindex! at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:126 [inlined]
 [15] MapUpdater at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:241 [inlined]
 [16] (::Observables.OnUpdate{Observables.MapUpdater{typeof(MakieLayout.ylimits),Tuple{Float32,Float32}},Tuple{Observable{GeometryBasics.HyperRectangle{2,Float32}}}})(::GeometryBasics.HyperRectangle{2,Float32}) at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:218
 [17] setindex!(::Observable{GeometryBasics.HyperRectangle{2,Float32}}, ::GeometryBasics.HyperRectangle{2,Float32}; notify::Observables.var"#6#8") at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:130
 [18] setindex!(::Observable{GeometryBasics.HyperRectangle{2,Float32}}, ::GeometryBasics.HyperRectangle{2,Float32}) at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:126
 [19] adjustlimits!(::MakieLayout.LAxis) at C:\Users\fbanning\.julia\packages\MakieLayout\3a66O\src\lobjects\laxis.jl:533
 [20] #163 at C:\Users\fbanning\.julia\packages\MakieLayout\3a66O\src\lobjects\laxis.jl:339 [inlined]
 [21] (::Observables.OnUpdate{MakieLayout.var"#163#187"{MakieLayout.LAxis},Tuple{Observable{GeometryBasics.HyperRectangle{2,Int64}},Observable{Any}}})(::GeometryBasics.HyperRectangle{2,Float32}) at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:218
 [22] setindex!(::Observable{Any}, ::GeometryBasics.HyperRectangle{2,Float32}; notify::Observables.var"#6#8") at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:130
 [23] setindex!(::Observable{Any}, ::GeometryBasics.HyperRectangle{2,Float32}) at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:126
 [24] autolimits!(::MakieLayout.LAxis) at C:\Users\fbanning\.julia\packages\MakieLayout\3a66O\src\lobjects\laxis.jl:514
 [25] plot!(::MakieLayout.LAxis, ::Type{Lines{...}}, ::Attributes, ::Observable{Array{Point{2,Float32},1}}; kw_attributes::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at C:\Users\fbanning\.julia\packages\MakieLayout\3a66O\src\lobjects\laxis.jl:352
 [26] plot! at C:\Users\fbanning\.julia\packages\MakieLayout\3a66O\src\lobjects\laxis.jl:350 [inlined]
 [27] #lines!#131 at C:\Users\fbanning\.julia\packages\AbstractPlotting\q9DyS\src\recipes.jl:23 [inlined]
 [28] vline!(::MakieLayout.LAxis, ::Int64; kwargs::Base.Iterators.Pairs{Symbol,String,Tuple{Symbol},NamedTuple{(:color,),Tuple{String}}}) at C:\Users\fbanning\.julia\packages\InteractiveChaos\m3no4\src\agents\agents.jl:280
 [29] add_reset_line!(::Array{Any,1}, ::Int64) at C:\Users\fbanning\.julia\packages\InteractiveChaos\m3no4\src\agents\agents.jl:285
 [30] (::InteractiveChaos.var"#31#41"{typeof(ac),Float64,Symbol,typeof(random_activation),Nothing,AgentBasedModel{MyAgent,ContinuousSpace{typeof(Agents.defvel),Tuple{Int64,Int64}},typeof(random_activation),Dict{Symbol,Real}},Observable{AgentBasedModel{MyAgent,ContinuousSpace{typeof(Agents.defvel),Tuple{Int64,Int64}},typeof(random_activation),Dict{Symbol,Real}}},Observable{Array{RGBA{Float32},1}},Float64,Symbol,Observable{Any}})(::Int64) at 
C:\Users\fbanning\.julia\packages\InteractiveChaos\m3no4\src\agents\agents.jl:153
 [31] #invokelatest#1 at .\essentials.jl:710 [inlined]
 [32] invokelatest at .\essentials.jl:709 [inlined]
 [33] setindex!(::Observable{Any}, ::Int64; notify::Observables.var"#6#8") at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:132
 [34] setindex!(::Observable{Any}, ::Int64) at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:126
 [35] (::MakieLayout.var"#281#289"{Observable{Any}})(::MakieLayout.MouseState{MakieLayout.MouseLeftClick}) at C:\Users\fbanning\.julia\packages\MakieLayout\3a66O\src\lobjects\lbutton.jl:80
 [36] (::MakieLayout.var"#71#72"{MakieLayout.var"#281#289"{Observable{Any}}})(::MakieLayout.MouseState{MakieLayout.MouseLeftClick}) at C:\Users\fbanning\.julia\packages\MakieLayout\3a66O\src\mousestatemachine.jl:45
 [37] #invokelatest#1 at .\essentials.jl:710 [inlined]
 [38] invokelatest at .\essentials.jl:709 [inlined]
 [39] setindex!(::Observable{MakieLayout.MouseState}, ::MakieLayout.MouseState{MakieLayout.MouseLeftClick}; notify::Observables.var"#6#8") at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:132
 [40] setindex!(::Observable{MakieLayout.MouseState}, ::MakieLayout.MouseState{MakieLayout.MouseLeftClick}) at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:126
 [41] (::MakieLayout.var"#85#87"{Scene,Module,Base.RefValue{Bool},Base.RefValue{Union{Nothing, AbstractPlotting.Mouse.Button}},Base.RefValue{Bool},Base.RefValue{Bool},Base.RefValue{Point{2,Float32}},Base.RefValue{Float64},Base.RefValue{Float64},Base.RefValue{Union{Nothing, AbstractPlotting.Mouse.Button}},Float64,Base.RefValue{Bool},Observable{MakieLayout.MouseState},MakieLayout.var"#is_mouse_over_relevant_area#86"{Scene,Tuple{Poly{...},Text{...}}}})(::Tuple{Float64,Float64}, ::AbstractPlotting.Mouse.DragEnum) at 
C:\Users\fbanning\.julia\packages\MakieLayout\3a66O\src\mousestatemachine.jl:205
 [42] (::Observables.OnUpdate{MakieLayout.var"#85#87"{Scene,Module,Base.RefValue{Bool},Base.RefValue{Union{Nothing, AbstractPlotting.Mouse.Button}},Base.RefValue{Bool},Base.RefValue{Bool},Base.RefValue{Point{2,Float32}},Base.RefValue{Float64},Base.RefValue{Float64},Base.RefValue{Union{Nothing, AbstractPlotting.Mouse.Button}},Float64,Base.RefValue{Bool},Observable{MakieLayout.MouseState},MakieLayout.var"#is_mouse_over_relevant_area#86"{Scene,Tuple{Poly{...},Text{...}}}},Tuple{Observable{Tuple{Float64,Float64}},Observable{AbstractPlotting.Mouse.DragEnum}}})(::AbstractPlotting.Mouse.DragEnum) at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:218
 [43] setindex!(::Observable{AbstractPlotting.Mouse.DragEnum}, ::AbstractPlotting.Mouse.DragEnum; notify::Observables.var"#6#8") at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:130        
 [44] setindex! at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:126 [inlined]
 [45] #776 at C:\Users\fbanning\.julia\packages\AbstractPlotting\q9DyS\src\interaction\events.jl:29 [inlined]
 [46] (::Observables.OnUpdate{AbstractPlotting.var"#776#777"{Base.RefValue{Bool},Base.RefValue{AbstractPlotting.Mouse.Button},Base.RefValue{AbstractPlotting.Mouse.DragEnum},Events},Tuple{Observable{Tuple{Float64,Float64}},Observable{Set{AbstractPlotting.Mouse.Button}}}})(::Set{AbstractPlotting.Mouse.Button}) at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:218
 [47] setindex!(::Observable{Set{AbstractPlotting.Mouse.Button}}, ::Set{AbstractPlotting.Mouse.Button}; notify::Observables.var"#6#8") at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:130  
 [48] setindex!(::Observable{Set{AbstractPlotting.Mouse.Button}}, ::Set{AbstractPlotting.Mouse.Button}) at C:\Users\fbanning\.julia\packages\Observables\0wrF6\src\Observables.jl:126
 [49] addbuttons(::Scene, ::Symbol, ::GLFW.MouseButton, ::GLFW.Action, ::Type{AbstractPlotting.Mouse.Button}) at C:\Users\fbanning\.julia\packages\GLMakie\wpJsD\src\events.jl:36
 [50] (::GLMakie.var"#mousebuttons#61"{Scene})(::GLFW.Window, ::GLFW.MouseButton, ::GLFW.Action, ::Int32) at C:\Users\fbanning\.julia\packages\GLMakie\wpJsD\src\events.jl:112
 [51] _MouseButtonCallbackWrapper(::GLFW.Window, ::GLFW.MouseButton, ::GLFW.Action, ::Int32) at C:\Users\fbanning\.julia\packages\GLFW\g1nX6\src\callback.jl:43
 [52] PollEvents at C:\Users\fbanning\.julia\packages\GLFW\g1nX6\src\glfw3.jl:585 [inlined]
 [53] renderloop(::GLMakie.Screen; framerate::Float64, prerender::GLMakie.var"#53#55") at C:\Users\fbanning\.julia\packages\GLMakie\wpJsD\src\rendering.jl:6
 [54] renderloop(::GLMakie.Screen) at C:\Users\fbanning\.julia\packages\GLMakie\wpJsD\src\rendering.jl:3
```

If this doesn't help you, I will cobble together a quick MWE (sorry, can't share the full code as part of it is research in progress).

**Comment by Datseris** [comment]
Hi, the problem is in MakieLayout unfortunately. I think the aggregated values provide some numbers they have trouble with. Here the MWE would be very helpful, because I need it to tell *them* the problem :)

**Comment by fbanning** [comment]
> I think the aggregated values provide some numbers they have trouble with

Could it be because the numbers in the model currently are pretty big? If so, why would that matter when I let it run a second time or when I press reset?

**Comment by Datseris** [comment]
I really cannot say more until I have some running code.

**Comment by fbanning** [comment]
I've tried to minimise it as far as reasonably possible.
Also added some comments for good measure. :)

Packages:
- Agents 3.3.1
- InteractiveChaos 0.9.0
- Makie 0.11.0

Steps to reproduce:
- Let the model `run` in Makie
- Click on `setup` a few times (sometimes it crashes after one click, sometimes it needs a few tries)

``` julia
using Agents
using InteractiveChaos, Makie

## Define agent type

mutable struct MyAgent <: AbstractAgent
    id::Int
    pos::Tuple{Float64, Float64}
    vel::Tuple{Float64, Float64}
    money::Float64
    transfer::Float64
end

## Define model functionality

function init_model(properties)
    space = ContinuousSpace(2; periodic = true, extend = (1, 1))
    model = ABM(MyAgent, space; scheduler= random_activation, properties = properties)
    init_agents(model, properties[:N])
    return model
end

function init_agents(model, N, id = 0)
    for _ in 1:N
        id += 1
        pos = Tuple(rand(2))
        vel = (0.0, 0.0)
        money = rand()
        agent= MyAgent(
            id, 
            pos, 
            vel, 
            money, 
            0
        )
        add_agent!(agent, model)
    end
    index!(model)
    return model
end

function agent_step!(agent, model)
    # randomise agent movement
    agent.vel = Tuple(rand(-0.001:0.0001:0.001, 2))
    # list of neighbors
    neighbors = space_neighbors(agent.pos, model, 0.02)
    # move if no neighbors
    if length(neighbors) ≤ 0
        move_agent!(agent, model)
    # share money if neighbors 
    else
        # take random neighbor as partner
        partner = model[rand(neighbors)]
        # determine amount of transfer
        agent.transfer = rand() * agent.money
        # reduce agent's money by amount of transfer
        agent.money -= agent.transfer
        # increase partner's money by amount of transfer times 3
        partner.money += agent.transfer * 3
        # determine amount of payback by partner
        payback = rand() * agent.transfer * 3
        # reduce partner's money by payback
        partner.money -= payback
        # increase agent's money by payback
        agent.money += payback
    end
end

## Initialise model

properties = Dict(
    :N => 100, # agent population
)

model = init_model(properties)

## Build interactive model

adata = [(:money, sum), (:transfer, sum)]
alabels = ["Aggregate \nmoney", "Aggregate \ntransfers"]

interactive_abm(
    model, agent_step!, dummystep; 
    as = 0.02,
    adata = adata, alabels = alabels
)
```

**Comment by fbanning** [comment]
Did you find the time to look into this issue? (No hurry, I'm just wondering.)

**Comment by Datseris** [comment]
Hi, nope but I will after #10 . Problem is, at the moment most things in InteractiveChaos don't work because of breaking changes due to the merging of MakieLayout. I'm waiting from support from the devs to continue further.

**Comment by Datseris** [comment]
A lot of work has gone into new versions here and in Agents, so I'm closing this now. If some issue persists, please open a new issue.
